mirgecom/simutil.py

```diff
@@ -162,6 +148,40 @@ def sim_checkpoint(discr, visualizer, eos, q, vizname, exact_soln=None,
         raise ExactSolutionMismatch(step, t=t, state=q)
 
 
+def sim_healthcheck(discr, eos, q, conserved_vars, step=0, t=0):
+    """Check the health of the simulation by checking for unphysical values."""
+    import grudge.op as op
+
+    # NOTE: Derived quantities are functions of the conserved variables.
+    # Therefore is it sufficient to check for unphysical values of
+    # temperature and pressure.
+    dependent_vars = eos.dependent_vars(conserved_vars)
+
+    # Check for NaN
+    if (np.isnan(op.nodal_sum(discr, "vol", dependent_vars.pressure))
+            or np.isnan(op.nodal_sum(discr, "vol", dependent_vars.temperature))):
```

Comment:
Need to synchronize the error condition across all ranks before raising, otherwise the code may hang. [Possibly useful snippet](https://github.com/illinois-ceesd/mirgecom/blob/e47d3d0b39b4782c05ef3d27d8cc63ceb84c47a3/mirgecom/mpi.py#L161-L166)

Reply:
Good point. Question though: Is this okay if `op.nodal_sum` and other reductions are distributed? Once https://github.com/inducer/grudge/pull/117 lands, these (and similar reductions like norms) will do parallel reductions. Or would we rather do things rank-local in these callbacks?

Reply:
Oh, neat. Yeah, I think if `op.nodal_sum` is global then it should be fine. It's probably going to be slower than doing a single reduction, but we can always optimize it later if it becomes a problem.

Reply:
@majosm I've opted to keep it rank-local for now, as exceptions are treated globally right? The error conditions now call out to rank-local grudge routines. Does this sound okay? I just felt doing 3 allreduces in a callback might be a bit much

Reply:
It is my understanding that exception handling *is not* global.  I think that's the main crux of our struggle with parallel/asynchronous error handling.  It appears to me, that if your current version of the healthcheck fails, then the simulation will hang or behave pathologically (as an exception gets tossed on one rank only, that rank goes into the handling branch, the rest don't).

I don't think we need a bunch of reductions though, really only one:

```python
def fluid_healthcheck(...):
   nans_present = check_for_local_naninf()
   if nans_present:
      message += "NANs or INFs present in solution"
   health_problem = check_for_whatever_local_problem(...):
   if healt_problem:
      message += "Some other health issue."
   health = nansyes | healthproblem | yadda | yadda | yadda
   health = comm.allreduce(health, MPI.LOR)
   if health:
      log(message)
      toss TheException(message)
```

We could probably do better with the actual message. Consider that we could have Nranks disparate messages, each with valid (and potentially useful) error content. If we use `log` or maybe a similar utility that we ourselves roll to be parallel-smart, then we can capture such messages.

Reply:
> It is my understanding that exception handling _is not_ global. I think that's the main crux of our struggle with parallel/asynchronous error handling. It appears to me, that if your current version of the healthcheck fails, then the simulation will hang or behave pathologically (as an exception gets tossed on one rank only, that rank goes into the handling branch, the rest don't).

Yeah. One minor clarification for @thomasgibson: if an exception is raised and doesn't get handled anywhere, it will bubble up to the top and cause mpi4py to call `MPI_Abort` (this is what `python -m mpi4py` does). In that situation it's fine for exceptions to occur locally. It's only when we try to handle them and do I/O, etc. prior to exiting that we need to synchronize globally.